crypto/evp/keymgmt_lib.c

```diff
@@ -183,6 +200,124 @@ void *evp_keymgmt_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
+    EVP_KEYMGMT *keymgmt2 = pkey2->pkeys[0].keymgmt;
+    void *provdata1 = NULL;
+    void *provdata2 = NULL;
+    void *exported_provdata2 = NULL;
+    int rv = 0;                  /* Assume the worst */
+
+    if (keymgmt1 == NULL && keymgmt2 == NULL)
+        return -2;   /* No support without at least one key manager */
+
+    if (keymgmt1 == NULL) {
+        EVP_KEYMGMT *swaptmp = keymgmt1;
+        keymgmt1 = keymgmt2;
+        keymgmt2 = swaptmp;
+    }
+
+    if (keymgmt1 != NULL && keymgmt2 != NULL
+        && EVP_KEYMGMT_number(keymgmt1) != EVP_KEYMGMT_number(keymgmt2))
+        return -1;               /* key type mismatch */
+
+    provdata1 = pkey1->pkeys[0].provdata;
+    provdata2 = pkey2->pkeys[0].provdata;
+
+    if (pkey1->pkeys[0].domainparams != domainparams
+        || pkey2->pkeys[0].domainparams != domainparams)
+        return -2;      /* No support for comparing stuff not asked */
+
+    /*
+     * If the two keys belong with different providers or the second key
+     * isn't provided, try to export the second provdata to the first
+     * provider and use that for comparing.
+     */
+    if (keymgmt1 != keymgmt2)
+        provdata2 = exported_provdata2 =
+            evp_keymgmt_export_to_provdata(pkey2, keymgmt1, domainparams);
+
+    /* TODO(3.0) Investigate if this should be -2, i.e unsupported */
+    if (provdata2 == NULL)
+        return -1;               /* Key type mismatch */
+
+    /*
+     * From here on, we will only return 0 or 1, because we know that
+     * the comparing functions are present in the EVP_KEYMGMT structure.
+     */
+    if (provdata1 != NULL && provdata2 != NULL)
+        rv = domainparams
+            ? evp_keymgmt_cmpdomparams(keymgmt1, provdata1, provdata2)
+            : evp_keymgmt_cmpkey(keymgmt1, provdata1, provdata2);
+
+    if (domainparams)
+        evp_keymgmt_freedomparams(keymgmt1, exported_provdata2);
+    else
+        evp_keymgmt_freekey(keymgmt1, exported_provdata2);
+
+    return rv;
+}
+
+int evp_keymgmt_copy(EVP_PKEY *to, const EVP_PKEY *from, int domainparams)
+{
+    EVP_KEYMGMT *keymgmt_to = to->pkeys[0].keymgmt;
+    EVP_KEYMGMT *keymgmt_from = from->pkeys[0].keymgmt;
+    void *provdata_to = NULL;
+
+    if (keymgmt_from == NULL || from->pkeys[0].domainparams != domainparams)
+        return 0;
+
+    if (keymgmt_to != NULL)
+        EVP_KEYMGMT_up_ref(keymgmt_to); /* ref++ */
+    evp_keymgmt_clear_pkey_cache(to);
+
+    if (keymgmt_to == NULL || keymgmt_to == keymgmt_from) {
+        if (keymgmt_to != NULL)
+            EVP_KEYMGMT_free(keymgmt_to); /* ref-- */
+        keymgmt_to = keymgmt_from;
+        EVP_KEYMGMT_up_ref(keymgmt_to); /* ref++ */
+        provdata_to = domainparams
+            ? evp_keymgmt_dupdomparams(keymgmt_from, from->pkeys[0].provdata, 1)
+            : evp_keymgmt_dupkey(keymgmt_from, from->pkeys[0].provdata, 1);
+    } else {
+        provdata_to =
+            evp_keymgmt_export_to_provdata(from, keymgmt_to, domainparams);
```

Comment:
Same comment as above. Why avoid the dirt count checking?

Reply:
Because in this case, we're not updating `from`, we're extracting provdata for `keymgmt_to`, based on `from`.